### PR TITLE
feat: add task executor trait

### PR DESCRIPTION
**Pull request by giladchase**
Proper usage example via tokio in upcoming PR, here is an abstract one:

```rust
 async fn initiates_a_cpu_bound_task() -> Result<(), Box<dyn std::error::Error>>{
    let executor = SomethingThatImplementsTaskExecutor::new();
    let derp = "DERP".to_string();
    let res = executor.spawn_blocking(move || {
	// Simulate blocking cpu computation
	let derp_world += derp + " world"
	// return the result of the computation
	derp_world
    }).await?;

    assert_eq!(res.as_str(), "DERP world");

    Ok(())
 }
```

```rust
 async fn initiates_an_io_bound_task() -> Result<(), Box<dyn std::error::Error>>{
    let executor = SomethingThatImplementsTaskExecutor::new();
    let derp = "DERP";
    let res = executor.spawn(async move || {
	tokio::time::sleep(tokio::time::Duration::from_secs(1)).await; // async sleep.
	println!("{derp}");
	derp
    }).await?;

    assert_eq!(res.as_str(), "DERP");

    Ok(())
}
```

commit-id:c6d2cb55

---

**Stack**:
- #51
- #50
- #49 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry, @elintul, and @giladchase)

*[a discussion](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVd_uF-JxybBVLcehe:-NvVd_uF-JxybBVLcehf:b-83sjoh) (no related file):*
Nice!

___
___
*[`crates/task_executor/src/executor.rs` line 6 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVZX_REF1nDJaRDKBZ:-NvVZX_REF1nDJaRDKB_:b-i9eut6) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/crates/task_executor/src/executor.rs#L6)):*
> ```XML
> /// A thread-safe, heap-allocated future, resolving to either success (`T`) or failure (`E`).
> /// Note: Pinned return type is in case we use self-referencing or other non-movable structs,
> pub(crate) type ExecutorFuture<T, E> = Pin<Box<dyn Future<Output = Result<T, E>> + Send>>;
> ```

Why?
We need to use it in `mempool` and `gateway` crates as well.



_Code quote:_
```XML
pub(crate)
```
___
*[`crates/task_executor/src/executor.rs` line 6 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVdMRj3vQCELFv3GbV:-NvVdMRj3vQCELFv3GbW:b-84bkng) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/crates/task_executor/src/executor.rs#L6)):*
> ```XML
> /// A thread-safe, heap-allocated future, resolving to either success (`T`) or failure (`E`).
> /// Note: Pinned return type is in case we use self-referencing or other non-movable structs,
> pub(crate) type ExecutorFuture<T, E> = Pin<Box<dyn Future<Output = Result<T, E>> + Send>>;
> ```

Could you explain why you use `Pin`?



_Code quote:_
```XML
Pin
```

**Review by MohammadNassar1 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 4 unresolved discussions (waiting on @dafnamatsry, @elintul, and @giladchase)

___
*[`Cargo.toml` line 25 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVduIXEVokHzsKoNIf:-NvVduIXEVokHzsKoNIg:ba3dfna) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/Cargo.toml#L25)):*
> ```TOML
> axum = "0.6.12"
> clap = { version = "4.3.10" }
> futures = "0.3.30"
> ```

Remove?



_Code quote:_
```TOML
futures = "0.3.30"
```

**Review by dafnamatsry — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: 4 of 5 files reviewed, 5 unresolved discussions (waiting on @elintul, @giladchase, and @MohammadNassar1)

___
*[`crates/task_executor/src/executor.rs` line 6 at r2](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvaRGrvEYkKTi83PQ8x:-NvaRGrvEYkKTi83PQ8y:b-6ewamb) ([raw file](https://github.com/starkware-libs/mempool/blob/3429e3f828075a2e07df854463f5e454a42325bb/crates/task_executor/src/executor.rs#L6)):*
> ```XML
> /// A thread-safe, heap-allocated future, resolving to either success (`T`) or failure (`E`).
> /// Note: Pinned return type is in case we use self-referencing or other non-movable structs,
> pub(crate) type ExecutorFuture<T, E> = Pin<Box<dyn Future<Output = Result<T, E>> + Send>>;
> ```

Have you considered using `BoxFuture`? (https://docs.rs/futures/latest/futures/future/type.BoxFuture.html)



_Code quote:_
```XML
Pin<Box<dyn Future<Output = Result<T, E>> + Send>>
```

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: 4 of 5 files reviewed, 5 unresolved discussions (waiting on @dafnamatsry, @elintul, and @MohammadNassar1)

*[a discussion](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVd_uF-JxybBVLcehe:-NvawkoM3rMCfR6y2M1J:b9gx4q) (no related file):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Nice!
</blockquote></details>

:pray:  :)

___
___
*[`Cargo.toml` line 25 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVduIXEVokHzsKoNIf:-Nv_zIU97QXEkBoTuxEI:b-896fix) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/Cargo.toml#L25)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Remove?
</blockquote></details>

Done.
___
*[`crates/task_executor/src/executor.rs` line 6 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVZX_REF1nDJaRDKBZ:-Nv_zSl56DGpY7VHhFr4:b-in71n6) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Why?
We need to use it in `mempool` and `gateway` crates as well.
</blockquote></details>

Right, tnx!
___
*[`crates/task_executor/src/executor.rs` line 6 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVdMRj3vQCELFv3GbV:-Nv_zWLk3fd7OfDnSHfZ:b-mwkrcb) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Could you explain why you use `Pin`?
</blockquote></details>

Sure, I'll explain briefly here and schedule a short meeting with you, Elin, Tsabari to discuss the whole thing anyway to ensure that all the technical details are valid (at Lior's request).

tl;dr
Pinning Futures is an implementation detail for `await`ing a Future.
I edited the docstring t reflect this more directly,

### Explanation: 

`Pin<Future>` prevents the `Future` memory address from being moved (for example via `std::mem::{swap, take}`).
 
Loosely speaking, this is an issue for `Future`s because when compiling a Future, the compiler occasionally generates self-referrential structs (like structs containing a field that is a [raw pointer](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html#why-pinning) to the address of another field in the same struct).

 Moving the address of such structs inherently breaks rust's safety assumptions because post-move, the pointer field would still point to the original address of the field (despite the value being moved from it!).

In other words, there are very few cases where you'll want non-pinned Futures around, due them not being awaitable.
___
*[`crates/task_executor/src/executor.rs` line 6 at r2](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvaRGrvEYkKTi83PQ8x:-NvatV71DUEgwDPFNruM:bqvnzc9) ([raw file](https://github.com/starkware-libs/mempool/blob/3429e3f828075a2e07df854463f5e454a42325bb/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Have you considered using `BoxFuture`? (https://docs.rs/futures/latest/futures/future/type.BoxFuture.html)
</blockquote></details>

Ya that's where I got it from, but I didn't use it because:

1. didn't want to add it as a (non-dev) dep just to get the typedef. 
2. it seems to be intended for use in weird lifetime edgecases (hence the lifetime param they add which we don't need), and I was worried that the name `BoxFuture` might mislead people who know it as such. From last paragraph [here](https://rust-lang.github.io/async-book/07_workarounds/04_recursion.html?highlight=BoxFuture#recursion) and first paragraph [here](https://docs.rs/futures/latest/futures/future/type.BoxFuture.html)

**Review by MohammadNassar1 — APPROVED** [review]
:lgtm:

> Reviewed 1 of 1 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49#-:-NvbWweq-fmpOjNWzGnd:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @dafnamatsry, @elintul, and @giladchase)

___
*[`crates/task_executor/src/executor.rs` line 6 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVdMRj3vQCELFv3GbV:-NvbWosb-sueCKJfqP9W:b-ucfknh) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Sure, I'll explain briefly here and schedule a short meeting with you, Elin, Tsabari to discuss the whole thing anyway to ensure that all the technical details are valid (at Lior's request).

tl;dr
Pinning Futures is an implementation detail for `await`ing a Future.
I edited the docstring t reflect this more directly,

### Explanation: 

`Pin<Future>` prevents the `Future` memory address from being moved (for example via `std::mem::{swap, take}`).
 
Loosely speaking, this is an issue for `Future`s because when compiling a Future, the compiler occasionally generates self-referrential structs (like structs containing a field that is a [raw pointer](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html#why-pinning) to the address of another field in the same struct).

 Moving the address of such structs inherently breaks rust's safety assumptions because post-move, the pointer field would still point to the original address of the field (despite the value being moved from it!).

In other words, there are very few cases where you'll want non-pinned Futures around, due them not being awaitable.
</blockquote></details>

Thanks!

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: 4 of 5 files reviewed, 2 unresolved discussions (waiting on @dafnamatsry, @elintul, and @MohammadNassar1)

___
*[`crates/task_executor/src/executor.rs` line 6 at r1](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvVdMRj3vQCELFv3GbV:-NvgkKkzC6DzdViwLZK6:b-y3dps0) ([raw file](https://github.com/starkware-libs/mempool/blob/bf9ffcb59f54d34a44a1c6dc4d33614f05ac0e78/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, MohammadNassar1 (mohammad-starkware) wrote…</i></summary><blockquote>

Thanks!
</blockquote></details>

No longer relevant anyway :sweat_smile: see my response to Dafna below
___
*[`crates/task_executor/src/executor.rs` line 6 at r2](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvaRGrvEYkKTi83PQ8x:-NvgkCsu9Eu2scbrDWFj:b24kjzh) ([raw file](https://github.com/starkware-libs/mempool/blob/3429e3f828075a2e07df854463f5e454a42325bb/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Ya that's where I got it from, but I didn't use it because:

1. didn't want to add it as a (non-dev) dep just to get the typedef. 
2. it seems to be intended for use in weird lifetime edgecases (hence the lifetime param they add which we don't need), and I was worried that the name `BoxFuture` might mislead people who know it as such. From last paragraph [here](https://rust-lang.github.io/async-book/07_workarounds/04_recursion.html?highlight=BoxFuture#recursion) and first paragraph [here](https://docs.rs/futures/latest/futures/future/type.BoxFuture.html)
</blockquote></details>

Ended up not needing Pin, tokio already handles Pinning under the hood when awaiting a future.

It works since [the method](https://doc.rust-lang.org/std/future/trait.Future.html) inside the `Future` trait already requires self to be `Pin`ed, so whatever engine that tries to call it will have to Pin it.

**Review by MohammadNassar1 — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 1 unresolved discussion (waiting on @elintul and @giladchase)

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 1 unresolved discussion (waiting on @dafnamatsry and @elintul)

___
*[`crates/task_executor/src/executor.rs` line 6 at r2](https://reviewable.io/reviews/starkware-libs/mempool/49#-NvaRGrvEYkKTi83PQ8x:-NvlPRde5tKKDGqpP43t:b-8dzjh0) ([raw file](https://github.com/starkware-libs/mempool/blob/3429e3f828075a2e07df854463f5e454a42325bb/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, dafnamatsry wrote…</i></summary><blockquote>

Cool. It makes more sense that we shouldn't manage these kind of things ourselves.
I found a very good blog post that explains the async/await in Rust (with futures, pinning, etc.) : https://os.phil-opp.com/async-await/
</blockquote></details>

Oooh nice one :star_struck: 
I'll TAL, tnx!

**Review by elintul — APPROVED** [review]
:lgtm:

> Reviewed 3 of 5 files at r1, 1 of 1 files at r2, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49#-:-Nwn_yj_23BqXDyxrAlM:bnfp4nl)* status: all files reviewed, 2 unresolved discussions (waiting on @dafnamatsry and @giladchase)

___
*[`crates/task_executor/src/executor.rs` line 20 at r5](https://reviewable.io/reviews/starkware-libs/mempool/49#-Nwn_h_T8XNfxX2r0o-1:-Nwn_h_T8XNfxX2r0o-2:undefined) ([raw file](https://github.com/starkware-libs/mempool/blob/90ea50033668086f302b0a5ad02aee4e9e570268/crates/task_executor/src/executor.rs#L20)):*
> ```XML
>     /// Offloads a non-blocking task asynchronously.
>     /// It accepts a future representing an asynchronous operation and returns a result.
>     fn spawn<F, T>(&self, task: F) -> impl Future<Output = Result<T, Self::SpawnError>> + Send
> ```





_Suggestion:_
```XML
JoinHandle<T, Self::SpawnError>
```

File: crates/task_executor/src/executor.rs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    where
+        F: FnOnce() -> T + Send + 'static,
+        T: Send + 'static;
+
+    /// Offloads a non-blocking task asynchronously.
+    /// It accepts a future representing an asynchronous operation and returns a result.
+    fn spawn<F, T>(
+        &self,
+        task: F,
+    ) -> impl Future<Output = Result<T, Self::SpawnBlockingError>> + Send
```
Comment (ittaysw) [inline crates/task_executor/src/executor.rs:23]: smol, for example, returns Task, will it be compatible ?

File: crates/task_executor/src/executor.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+use std::future::Future;
+
+/// An abstraction for executing tasks, suitable for both CPU-bound and I/O-bound operations.
+pub trait TaskExecutor {
+    type SpawnBlockingError;
+    type SpawnError;
```
Comment (ittaysw) [inline crates/task_executor/src/executor.rs:6]: When we have client code that will be specific to the concrete types of these errors, it may not be trivial to change with another implementation.

**Review by ittaysw — APPROVED** [review]
.

**Review by elintul — APPROVED** [review]
> Reviewed 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @giladchase)

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @ittaysw)

___
*[`crates/task_executor/src/executor.rs` line 6 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626925:-NyGWhXIEn4Hd594nudb:b-kzpqvz) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, ittaysw (Ittay Dror) wrote…</i></summary><blockquote>

When we have client code that will be specific to the concrete types of these errors, it may not be trivial to change with another implementation. 
</blockquote></details>

Good point, how about trait-bounding these error types with a dedicated trait? Something that will standardize the usage of them?
___
*[`crates/task_executor/src/executor.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626915:-NyGLGzM8O6iTSuS2q3o:b-93kkin) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L23)):*
<details><summary><i>Previously, ittaysw (Ittay Dror) wrote…</i></summary><blockquote>

smol, for example, returns Task, will it be compatible ? 
</blockquote></details>

Ya, but i think we'll need to add `catch_unwind` ourselves and map its error to a new error enum.
Something like:
```rust
    fn spawn<F, T>(&self, task: F) -> impl Future<Output = Result<T, Self::SpawnError>> + Send
    where
        F: Future<Output = T> + Send + 'static,
        T: Send + 'static,
    {
        smol::spawn(async move {
            std::panic::AssertUnwindSafe(task)
                .catch_unwind()
                .await
                .map_err(|_| SmolExecutorError::PanicError)
        })
    }

enum SmolExecutorError {...}
```

This is because I think only tokio catches panics out of the box (so perhaps this interface is a bit tokio-biased but it might not necessarily be a bad thing).

Reasons why I think smol doesn't catch panics:
1) tried it and it didn't catch unless I added catch_unwind myself :p
2) their code only has catch_unwind [here](https://github.com/smol-rs/smol/blob/641326b54f63b017758f2bfe85dfb3662ed51283/src/spawn.rs#L2), which seems to be for internal usage.
3) [this](https://github.com/smol-rs/smol/issues/202) old issue has maintainers saying they prefer propagating panics and let  users decide if they should be caught or not.

**Review by elintul — COMMENTED** [review]
> Reviewed 2 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: 4 of 5 files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @ittaysw)

**Review by elintul — APPROVED** [review]
> Reviewed 1 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @ittaysw)

**Review by ittaysw — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @giladchase)

___
*[`crates/task_executor/src/executor.rs` line 6 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626925:-NyOl1opB4bsazqJD0pc:b-31oij0) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Good point, how about trait-bounding these error types with a dedicated trait? Something that will standardize the usage of them?
</blockquote></details>

If you mean having a SpawnError trait, then it means you have two options: (1) wrap the concrete error with this type but then if the client code needs to react to it, it'll need to unwrap to figure out what the real error was, making it depend on the concrete error again, or (2) have the SpawnError cover all possible errors in the underlying implementation but then you need to look at several implementations to make sure you cover all bases
___
*[`crates/task_executor/src/executor.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626915:-NyOlss1-LlgLPaeqG1A:btktk2o) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L23)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Ya, but i think we'll need to add `catch_unwind` ourselves and map its error to a new error enum.
Something like:
```rust
    fn spawn<F, T>(&self, task: F) -> impl Future<Output = Result<T, Self::SpawnError>> + Send
    where
        F: Future<Output = T> + Send + 'static,
        T: Send + 'static,
    {
        smol::spawn(async move {
            std::panic::AssertUnwindSafe(task)
                .catch_unwind()
                .await
                .map_err(|_| SmolExecutorError::PanicError)
        })
    }

enum SmolExecutorError {...}
```

This is because I think only tokio catches panics out of the box (so perhaps this interface is a bit tokio-biased but it might not necessarily be a bad thing).

Reasons why I think smol doesn't catch panics:
1) tried it and it didn't catch unless I added catch_unwind myself :p
2) their code only has catch_unwind [here](https://github.com/smol-rs/smol/blob/641326b54f63b017758f2bfe85dfb3662ed51283/src/spawn.rs#L2), which seems to be for internal usage.
3) [this](https://github.com/smol-rs/smol/issues/202) old issue has maintainers saying they prefer propagating panics and let  users decide if they should be caught or not.
</blockquote></details>

If the idea of the interface is to have two implementations: tokio or something trivial for testing (that is, without a scheduler, multi threading, etc.), then I'm OK with it. If it tries to be generic so that in the future we can replace tokio with smol or some other implementation, then we should be more careful

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @ittaysw)

___
*[`crates/task_executor/src/executor.rs` line 6 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626925:-NyPCWcI5WnBUYxbuhHC:b-o41c6q) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, ittaysw (Ittay Dror) wrote…</i></summary><blockquote>

If you mean having a SpawnError trait, then it means you have two options: (1) wrap the concrete error with this type but then if the client code needs to react to it, it'll need to unwrap to figure out what the real error was, making it depend on the concrete error again, or (2) have the SpawnError cover all possible errors in the underlying implementation but then you need to look at several implementations to make sure you cover all bases
</blockquote></details>

I see your point :thinking: 

How about going the other direction then? 
Let's remove the associated types and add a concrete error type in this module that all implementations have to `map_err` into.

Error variants of this concrete type could be: ThreadPanicked, ExecutorResourcesExhaustion, TaskTimeout, TaskAborted, others?
___
*[`crates/task_executor/src/executor.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626915:-NyPEKey3lD1nvIdkiN-:blnrahf) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L23)):*
<details><summary><i>Previously, ittaysw (Ittay Dror) wrote…</i></summary><blockquote>

If the idea of the interface is to have two implementations: tokio or something trivial for testing (that is, without a scheduler, multi threading, etc.), then I'm OK with it. If it tries to be generic so that in the future we can replace tokio with smol or some other implementation, then we should be more careful
</blockquote></details>

I think option 2 was what we ended up with when we met, so let's be careful :)

What do you think we need to change in order to comply with this? 
For example, we remove the `Result` from the API's return types, which will work for both `smol` and `Tokio` (where in tokio the type of T will just be a result i guess?).
However this will effectively limit the executor's ability to catch panics, so I think I might be missing your point here :sweat_smile:

**Review by ittaysw — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @giladchase)

___
*[`crates/task_executor/src/executor.rs` line 6 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626925:-NyPPs4D8KfdUkGtXZyl:b1bcyht) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

I see your point :thinking: 

How about going the other direction then? 
Let's remove the associated types and add a concrete error type in this module that all implementations have to `map_err` into.

Error variants of this concrete type could be: ThreadPanicked, ExecutorResourcesExhaustion, TaskTimeout, TaskAborted, others?
</blockquote></details>

The 'others' may be the issue. And I don't know about the former. I think creating an "air tight" abstraction will be difficult. So I'm OK with not trying to much as long as we understand that replacing tokio with another (real) implementation may be hard. 
___
*[`crates/task_executor/src/executor.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626915:-NyPQCx03IcY9YPCWQwU:bixn9wh) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L23)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

I think option 2 was what we ended up with when we met, so let's be careful :)

What do you think we need to change in order to comply with this? 
For example, we remove the `Result` from the API's return types, which will work for both `smol` and `Tokio` (where in tokio the type of T will just be a result i guess?).
However this will effectively limit the executor's ability to catch panics, so I think I might be missing your point here :sweat_smile:
</blockquote></details>

I think option 2 is hard to do correctly and even if we start out correctly, others may inadvertently add tokio dependencies. E.g. the code example in Commits uses  tokio::time::sleep. 

So IMO, it is OK to have option 1

**Review by giladchase — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry and @ittaysw)

___
*[`crates/task_executor/src/executor.rs` line 6 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626925:-NyQ4iVR6i1E9TTMCTox:bbo6o9y) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, ittaysw (Ittay Dror) wrote…</i></summary><blockquote>

The 'others' may be the issue. And I don't know about the former. I think creating an "air tight" abstraction will be difficult. So I'm OK with not trying to much as long as we understand that replacing tokio with another (real) implementation may be hard.
</blockquote></details>

Got it, i'm cool with setting up a future challenge anyway :smirk: 
So I'm leaving it as is?
___
*[`crates/task_executor/src/executor.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626915:-NyQ4wlQ9AfwoefHWEhA:bhiq26v) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L23)):*
<details><summary><i>Previously, ittaysw (Ittay Dror) wrote…</i></summary><blockquote>

I think option 2 is hard to do correctly and even if we start out correctly, others may inadvertently add tokio dependencies. E.g. the code example in Commits uses  tokio::time::sleep. 

So IMO, it is OK to have option 1
</blockquote></details>

Option 1 it is :muscle: 
So no changes necessary?

**Review by ittaysw — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49#-:-NyTVigdCBs4lwIdlLqb:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @dafnamatsry)

**Review by giladchase — COMMENTED** [review]
> Dismissed @ittaysw from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: 2 of 5 files reviewed, 1 unresolved discussion (waiting on @dafnamatsry and @elintul)

___
*[`crates/task_executor/src/executor.rs` line 6 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626925:-NyU6I0UEFHkC-mipfYM:b-7pzdb4) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L6)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Got it, i'm cool with setting up a future challenge anyway :smirk: 
So I'm leaving it as is?
</blockquote></details>

Resolved offline
___
*[`crates/task_executor/src/executor.rs` line 23 at r6](https://reviewable.io/reviews/starkware-libs/mempool/49#gh-1597626915:-NyU6DTz3Enns2XmhgW-:b-qa4alu) ([raw file](https://github.com/starkware-libs/mempool/blob/842b997835607ab4e6eb8746e2a5a602089f2480/crates/task_executor/src/executor.rs#L23)):*
<details><summary><i>Previously, giladchase wrote…</i></summary><blockquote>

Option 1 it is :muscle: 
So no changes necessary?
</blockquote></details>

Resolved offline.

**Review by dafnamatsry — APPROVED** [review]
> Reviewed 1 of 3 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @giladchase)

**Review by giladchase — COMMENTED** [review]
> Reviewed 1 of 5 files at r1, 1 of 3 files at r7, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/starkware-libs/mempool/49)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @giladchase)
